# 🧪 Step 1: Research & Data Source Discovery

**PR Branch Name: supermarkets-data-modelling**

This notebook documents the process for Step 1 of the "Supermarkets in Berlin" project:

- 1.1 Data Source Discovery
- 1.2 Modelling & Planning
- 1.3 Prepare the /sources Directory
- 1.4 Review

**Goal:**

- Identify and document relevant data sources.
- Select key parameters for our use case.
- Draft the planned table schema.
- Plan cleaning and transformation steps before database population.

## 1.1 Data Source Discovery

Topic: Supermarkets in Berlin

**Main source:**

- Name: OpenStreetMap (OSM) via OSMnx library
- Source and origin: Public crowdsourced geospatial database
- Update frequency: Continuous (dynamic)
- Data type: Dynamic (API query using shop=supermarket)

**Reason for selection:**

- Covers all supermarkets in Berlin
- Includes coordinates, names, addresses, and other useful attributes
- Open, free, and easy to query programmatically

**Optional additional sources:**

- Name: Berlin Open Data Portal (daten.berlin.de)
- Source and origin: Official Berlin city government
- Update frequency: Varies per dataset
- Data type: Static or semi-static (download as CSV/GeoJSON)
- Possible usage: Enrich with official administrative boundaries or extra metadata

**Enrichment potential:**

Neighborhood/district info from Berlin shapefiles (GeoJSON)
Linking to local amenities for spatial context

**📍 Fetch data about supermarkets in Berlin from OpenStreetMap (OSM)**

In [37]:
# install libraries
# %pip install osmnx geopandas pandas

In [38]:
# import libraries
import osmnx as ox
import geopandas as gpd
import pandas as pd

In [39]:
# Give me all places tagged as shop=supermarket.
# tags filter for only features with 

tags = {"shop": "supermarket"}

In [40]:
# ✅ Enables caching: Speeds up repeated queries
# 🖥 Logs details to the console (helpful for debugging)
ox.settings.use_cache = True
ox.settings.log_console = True

In [41]:
# Fetch Supermarkets from Berlin from OSM using the tag "shop=supermarket"
gdf = ox.features.features_from_place("Berlin, Germany", tags=tags)

This line queries the OSM Overpass API and returns a GeoDataFrame (gdf) with all supermarkets in Berlin, including their geometry (coordinates) and OSM metadata (like name, address, brand, etc.).

In [42]:
# Display basic info

print(f"Number of supermarkets entries fetched: {len(gdf)}")
gdf.head(3)

Number of supermarkets entries fetched: 1358


geometry addr:city addr:country  \
element id                                                           
node    58489979  POINT (13.40737 52.50982)    Berlin           DE   
        79418658  POINT (13.30855 52.57242)    Berlin           DE   
        79422426  POINT (13.31248 52.57168)       NaN          NaN   

                 addr:housenumber addr:postcode       addr:street  \
element id                                                          
node    58489979               83         10179  Alte Jakobstraße   
        79418658                2         13403      Quäkerstraße   
        79422426              NaN           NaN               NaN   

                    addr:suburb                  brand brand:wikidata  \
element id                                                              
node    58489979          Mitte  Netto Marken-Discount        Q879858   
        79418658  Reinickendorf                    NaN            NaN   
        79422426            NaN                Nahkauf      Q57515238   

                           brand:wikipedia  ...  \
element id                                  ...   
node    58489979  de:Netto Marken-Discount  ...   
        79418658                       NaN  ...   
        79422426                       NaN  ...   

                 cash_withdrawal:purchase_minimum internet payment:ec  \
element id                                                              
node    58489979                              NaN      NaN        NaN   
        79418658                              NaN      NaN        NaN   
        79422426                              NaN      NaN        NaN   

                 disused:shop building:part  fee building:parts  \
element id                                                        
node    58489979          NaN           NaN  NaN            NaN   
        79418658          NaN           NaN  NaN            NaN   
        79422426          NaN           NaN  NaN            NaN   

                 safety:hand_sanitizer:covid19 operator:legal opening_date  
element id                                                                  
node    58489979                           NaN            NaN          NaN  
        79418658                           NaN            NaN          NaN  
        79422426                           NaN            NaN          NaN  

[3 rows x 230 columns]

In [43]:
#gdf.to_file("../sources/raw_supermarkets.geojson", driver="GeoJSON")

In [44]:
#gdf.to_csv("../sources/raw_supermarkets.csv", index=False)

In [45]:
gdf.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
MultiIndex: 1358 entries, ('node', np.int64(58489979)) to ('way', np.int64(1429504461))
Columns: 230 entries, geometry to opening_date
dtypes: geometry(1), object(229)
memory usage: 2.4+ MB


In [46]:
print(gdf.columns.tolist())


['geometry', 'addr:city', 'addr:country', 'addr:housenumber', 'addr:postcode', 'addr:street', 'addr:suburb', 'brand', 'brand:wikidata', 'brand:wikipedia', 'check_date', 'check_date:opening_hours', 'internet_access', 'name', 'opening_hours', 'payment:mastercard', 'payment:visa', 'shop', 'website', 'wheelchair', 'operator', 'phone', 'layer', 'payment:cards', 'payment:cash', 'payment:contactless', 'atm', 'atm:operator', 'type', 'wheelchair:description', 'contact:website', 'toilets:wheelchair', 'name:fa', 'diet:halal', 'diet:kosher', 'origin', 'diet:gluten_free', 'diet:lactose_free', 'diet:sugar_free', 'organic', 'contact:phone', 'operator:wikidata', 'internet_access:fee', 'level', 'payment:credit_cards', 'payment:girocard', 'email', 'branch', 'source', 'building', 'air_conditioning', 'drink:club-mate', 'payment:debit_cards', 'ref', 'self_checkout', 'stroller', 'toilets', 'payment:maestro', 'brand:website', 'addr:floor', 'diet:vegan', 'diet:vegetarian', 'name:de', 'note', 'currency:EUR', '

In [47]:
gdf['operator'].value_counts()

operator
Dennree GmbH                                  50
ALDI Nord                                     33
Edeka                                         21
Netto ApS & Co. KG / Dansk Supermarked A/S    17
Lidl                                          13
                                              ..
Bianca Unger                                   1
Biotopia GmbH                                  1
Bio Company GmbH                               1
Kaufland Vertrieb 228 GmbH & Co. KG            1
Ledo Goebenstr GmbH                            1
Name: count, Length: 185, dtype: int64

In [48]:
# filter columns that relevant to project
columns = [
    "name", "addr:street", "addr:housenumber", "addr:postcode",
    "addr:city", "opening_hours", "brand", "type","geometry","payment:credit_cards",
    "payment:debit_cards", "payment:cash", "payment:contactless", "wheelchair", "internet_access","layer","phone","contact:email","website"

]
gdf_superstore = gdf[[col for col in columns if col in gdf.columns]].copy()

In [49]:
gdf_superstore.head(3)

name       addr:street addr:housenumber  \
element id                                                                   
node    58489979  Netto Marken-Discount  Alte Jakobstraße               83   
        79418658                   Ledo      Quäkerstraße                2   
        79422426              kiezmarkt               NaN              NaN   

                 addr:postcode addr:city                      opening_hours  \
element id                                                                    
node    58489979         10179    Berlin          Mo-Sa 07:00-24:00; Su off   
        79418658         13403    Berlin                  Mo-Sa 09:00-20:00   
        79422426           NaN       NaN  Mo-Fr 08:00-20:00, Sa 08:00-19:00   

                                  brand type                   geometry  \
element id                                                                
node    58489979  Netto Marken-Discount  NaN  POINT (13.40737 52.50982)   
        79418658                    NaN  NaN  POINT (13.30855 52.57242)   
        79422426                Nahkauf  NaN  POINT (13.31248 52.57168)   

                 payment:credit_cards payment:debit_cards payment:cash  \
element id                                                               
node    58489979                  NaN                 NaN          NaN   
        79418658                  NaN                 NaN          NaN   
        79422426                  NaN                 NaN          NaN   

                 payment:contactless wheelchair internet_access layer phone  \
element id                                                                    
node    58489979                 NaN         no              no   NaN   NaN   
        79418658                 NaN        yes             NaN   NaN   NaN   
        79422426                 NaN        yes             NaN   NaN   NaN   

                 contact:email  \
element id                       
node    58489979           NaN   
        79418658           NaN   
        79422426           NaN   

                                                            website  
element id                                                           
node    58489979  https://www.netto-online.de/filialen/Berlin-Mi...  
        79418658                    https://www.ledo-supermarkt.de/  
        79422426                                                NaN

In [50]:
print('filtered dataset shape:', gdf_superstore.shape)
gdf_superstore.info()

filtered dataset shape: (1358, 19)
<class 'geopandas.geodataframe.GeoDataFrame'>
MultiIndex: 1358 entries, ('node', np.int64(58489979)) to ('way', np.int64(1429504461))
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype   
---  ------                --------------  -----   
 0   name                  1355 non-null   object  
 1   addr:street           1105 non-null   object  
 2   addr:housenumber      1097 non-null   object  
 3   addr:postcode         1081 non-null   object  
 4   addr:city             1044 non-null   object  
 5   opening_hours         1260 non-null   object  
 6   brand                 1134 non-null   object  
 7   type                  7 non-null      object  
 8   geometry              1358 non-null   geometry
 9   payment:credit_cards  152 non-null    object  
 10  payment:debit_cards   145 non-null    object  
 11  payment:cash          169 non-null    object  
 12  payment:contactless   37 non-null     object  
 13  wheelchair   

In [51]:
# Extract Latitude and Longitude
gdf_superstore['geometry'] = gdf_superstore['geometry'].apply(lambda geom: geom if geom.geom_type == 'Point' else geom.representative_point())
#Extract latitude and longitude

gdf_superstore["latitude"] = gdf.geometry.centroid.y
gdf_superstore["longitude"] = gdf.geometry.centroid.x

/var/folders/hj/bjtkr_b13175_g6c53_yly7c0000gn/T/ipykernel_61531/2522826528.py:5: UserWarning: Geometry is in a geographic CRS. Results from 'centroid' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

  gdf_superstore["latitude"] = gdf.geometry.centroid.y
/var/folders/hj/bjtkr_b13175_g6c53_yly7c0000gn/T/ipykernel_61531/2522826528.py:6: UserWarning: Geometry is in a geographic CRS. Results from 'centroid' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

  gdf_superstore["longitude"] = gdf.geometry.centroid.x


In [52]:
# rename columns for better understanding
gdf_superstore = gdf_superstore.rename(columns={
    "addr:street": "street",
    "addr:housenumber": "housenumber",
    "addr:postcode": "postcode",
    "addr:city": "city",
    "payment:credit_cards": "payment_credit_card",
    "payment:debit_cards": "payment_debit_cards",
    "payment:cash": "payment_cash",
    "payment:contactless": "payment_contactless",
    "contact:email": "email"
    })

In [53]:
gdf_superstore.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
MultiIndex: 1358 entries, ('node', np.int64(58489979)) to ('way', np.int64(1429504461))
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   name                 1355 non-null   object  
 1   street               1105 non-null   object  
 2   housenumber          1097 non-null   object  
 3   postcode             1081 non-null   object  
 4   city                 1044 non-null   object  
 5   opening_hours        1260 non-null   object  
 6   brand                1134 non-null   object  
 7   type                 7 non-null      object  
 8   geometry             1358 non-null   geometry
 9   payment_credit_card  152 non-null    object  
 10  payment_debit_cards  145 non-null    object  
 11  payment_cash         169 non-null    object  
 12  payment_contactless  37 non-null     object  
 13  wheelchair           1213 non-null   object  
 14  internet_acc

**⚠️ Data Quality Summary – Berlin Supermarkets Dataset**

- ✅ The dataset contains 1358 entries with latitude and longitude for all locations.

**📉 Missing Values**

- Address-related fields have some missing values:
- street (18% missing)
- housenumber (19% missing)
- postcode (20% missing)
- city (23% missing)
- Payment method availability is very sparse:
- payment_credit_card (only 11% present)
- payment_contactless (only 2.7% present)
- internet_access and type are also highly incomplete.
- brand is missing in ~16% of records.

- Accessibility: wheelchair accessibility is available in ~89% of stores – relatively well-covered.

⸻

**💡 Recommendation**

- Consider filling missing address values using reverse geocoding (based on lat/lon).
- Treat payment method columns as optional metadata.
- Drop or ignore type, layer, and other sparse columns unless specifically needed.


In [54]:
gdf_superstore.head()

name             street housenumber  \
element id                                                               
node    58489979  Netto Marken-Discount   Alte Jakobstraße          83   
        79418658                   Ledo       Quäkerstraße           2   
        79422426              kiezmarkt                NaN         NaN   
        79428988              nah & gut  Scharnweberstraße         100   
        79438509                Nahkauf       Meller Bogen           2   

                 postcode    city                      opening_hours  \
element id                                                             
node    58489979    10179  Berlin          Mo-Sa 07:00-24:00; Su off   
        79418658    13403  Berlin                  Mo-Sa 09:00-20:00   
        79422426      NaN     NaN  Mo-Fr 08:00-20:00, Sa 08:00-19:00   
        79428988    13405  Berlin                  Mo-Sa 07:00-22:00   
        79438509    13403  Berlin                  Mo-Sa 07:00-20:00   

                                  brand type                   geometry  \
element id                                                                
node    58489979  Netto Marken-Discount  NaN  POINT (13.40737 52.50982)   
        79418658                    NaN  NaN  POINT (13.30855 52.57242)   
        79422426                Nahkauf  NaN  POINT (13.31248 52.57168)   
        79428988                  EDEKA  NaN   POINT (13.31693 52.5664)   
        79438509                Nahkauf  NaN  POINT (13.32114 52.57049)   

                 payment_credit_card  ... payment_cash payment_contactless  \
element id                            ...                                    
node    58489979                 NaN  ...          NaN                 NaN   
        79418658                 NaN  ...          NaN                 NaN   
        79422426                 NaN  ...          NaN                 NaN   
        79428988                 NaN  ...          NaN                 NaN   
        79438509                 NaN  ...          yes                 yes   

                 wheelchair internet_access layer            phone email  \
element id                                                                 
node    58489979         no              no   NaN              NaN   NaN   
        79418658        yes             NaN   NaN              NaN   NaN   
        79422426        yes             NaN   NaN              NaN   NaN   
        79428988        yes             NaN   NaN  +49 30 41199655   NaN   
        79438509    limited             NaN     2    030-346491610   NaN   

                                                            website  \
element id                                                            
node    58489979  https://www.netto-online.de/filialen/Berlin-Mi...   
        79418658                    https://www.ledo-supermarkt.de/   
        79422426                                                NaN   
        79428988                                                NaN   
        79438509  https://www.nahkauf.de/maerkte/berlin-42613025...   

                   latitude  longitude  
element id                              
node    58489979  52.509819  13.407373  
        79418658  52.572417  13.308547  
        79422426  52.571676  13.312476  
        79428988  52.566403  13.316931  
        79438509  52.570486  13.321144  

[5 rows x 21 columns]

## 1.2 Modelling & Planning

**Selected 25 Key Columns**

 1.  osm_id 
 2.  name                 
 3.  street               
 4.  housenumber           
 5.  postcode              
 6.  city                   
 7.  opening_hours         
 8.  brand                  
 9.  type                  
 10. geometry             
 11. payment_credit_card   
 12. payment_debit_cards   
 13. payment_cash          
 14. payment_contactless    
 15. wheelchair            
 16. internet_access      
 17. layer                
 18. latitude           
 19. longitude 
 20. neighbourhood
 21. district
 22. source
 23. email
 24. phone
 25. website

**How this connects to existing tables:**

- Coordinates (latitude, longitude, geom): link to neighbourhood and district polygons.
- Neighbourhood & district fields: join with administrative boundaries table.
- Source field: ensures traceability.

### 🏪 Planned Schema: `superstore_in_berlin`

| Column Name             | Data Type | Description                                      | Example                            |
|-------------------------|-----------|--------------------------------------------------|------------------------------------|
| `osm_id`                | int       | Unique OSM element ID                            | 58489979                           |
| `name`                  | text      | Supermarket or store name                        | Netto Marken-Discount              |
| `brand`                 | text      | Brand if available                               | Nahkauf                            |
| `street`                | text      | Street name                                      | Alte Jakobstraße                   | 
| `housenumber`           | text      | House number                                     | 83                                 |
| `postcode`              | text      | Postal code                                      | 10179                              |
| `city`                  | text      | City name                                        | Berlin                             |
| `opening_hours`         | text      | Opening hours string                             | Mo–Sa 07:00–24:00; Su off          |
| `type`                  | text      | Store type if tagged                             | supermarket                        |
| `payment_credit_card`   | text      | Accepts credit card                              | yes                                |
| `payment_debit_cards`   | text      | Accepts debit cards                              | yes                                |
| `payment_cash`          | text      | Accepts cash payment                             | yes                                |
| `payment_contactless`   | text      | Accepts contactless payments                     | no                                 |
| `wheelchair`            | text      | Accessibility info                               | yes                                |
| `internet_access`       | text      | Public internet access (e.g., wifi)              | wlan                               |
| `layer`                 | text      | Vertical layer (e.g., floor number)              | 1                                  |
| `latitude`              | float     | Latitude coordinate                              | 52.5200                            |
| `longitude`             | float     | Longitude coordinate                             | 13.4050                            |
| `geometry`              | geometry  | Full GeoJSON geometry                            | POINT (13.4050 52.5200)            |
| `neighbourhood`         | text      | Local neighborhood (optional / derived)          | Kreuzberg                          |
| `district`              | text      | Berlin administrative district                   | Friedrichshain-Kreuzberg           |
| `source`                | text      | Data source info                                 | OSM                                |
| `email`                 | text      | Official store email                             | info@rewe-berlin.de                |
| `phone`                 | text      | Official store phone number                      | +49 30 12345678.                   |
| `website`               | text      | Official website                                 | https://www.lidl.de/filialen/berlin|

**Transformation Plan**

- Fetch data from OSM with filter shop=supershop (Berlin bounding box). ✅
- Clean column names → "addr:street" : "street".✅
- Normalize formats (Consider filling missing address values using reverse geocoding (based on lat/lon) ).📌

- Enrich with neighbourhood/district via spatial join. 📌

- Save cleaned dataset (GeoJSON + CSV).📌


In [55]:
gdf_superstore["source"]=None

In [56]:
gdf_superstore.head(3)

name            street housenumber  \
element id                                                              
node    58489979  Netto Marken-Discount  Alte Jakobstraße          83   
        79418658                   Ledo      Quäkerstraße           2   
        79422426              kiezmarkt               NaN         NaN   

                 postcode    city                      opening_hours  \
element id                                                             
node    58489979    10179  Berlin          Mo-Sa 07:00-24:00; Su off   
        79418658    13403  Berlin                  Mo-Sa 09:00-20:00   
        79422426      NaN     NaN  Mo-Fr 08:00-20:00, Sa 08:00-19:00   

                                  brand type                   geometry  \
element id                                                                
node    58489979  Netto Marken-Discount  NaN  POINT (13.40737 52.50982)   
        79418658                    NaN  NaN  POINT (13.30855 52.57242)   
        79422426                Nahkauf  NaN  POINT (13.31248 52.57168)   

                 payment_credit_card  ... payment_contactless wheelchair  \
element id                            ...                                  
node    58489979                 NaN  ...                 NaN         no   
        79418658                 NaN  ...                 NaN        yes   
        79422426                 NaN  ...                 NaN        yes   

                 internet_access layer phone email  \
element id                                           
node    58489979              no   NaN   NaN   NaN   
        79418658             NaN   NaN   NaN   NaN   
        79422426             NaN   NaN   NaN   NaN   

                                                            website  \
element id                                                            
node    58489979  https://www.netto-online.de/filialen/Berlin-Mi...   
        79418658                    https://www.ledo-supermarkt.de/   
        79422426                                                NaN   

                   latitude  longitude  source  
element id                                      
node    58489979  52.509819  13.407373    None  
        79418658  52.572417  13.308547    None  
        79422426  52.571676  13.312476    None  

[3 rows x 22 columns]

# Step 1 Review and A–F Data Familiarization

## A) Quick overview

In [57]:
print("Rows, Columns:", gdf_superstore.shape)
print("\nColumns:", gdf_superstore.columns.tolist())
print("data Info\n")
print(gdf_superstore.info())

Rows, Columns: (1358, 22)

Columns: ['name', 'street', 'housenumber', 'postcode', 'city', 'opening_hours', 'brand', 'type', 'geometry', 'payment_credit_card', 'payment_debit_cards', 'payment_cash', 'payment_contactless', 'wheelchair', 'internet_access', 'layer', 'phone', 'email', 'website', 'latitude', 'longitude', 'source']
data Info

<class 'geopandas.geodataframe.GeoDataFrame'>
MultiIndex: 1358 entries, ('node', np.int64(58489979)) to ('way', np.int64(1429504461))
Data columns (total 22 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   name                 1355 non-null   object  
 1   street               1105 non-null   object  
 2   housenumber          1097 non-null   object  
 3   postcode             1081 non-null   object  
 4   city                 1044 non-null   object  
 5   opening_hours        1260 non-null   object  
 6   brand                1134 non-null   object  
 7   type                 7 non-null

## B) Missing values per column

In [58]:
missing_count = gdf_superstore.isna().sum().sort_values(ascending=False)
print(missing_count)

source                 1358
layer                  1355
type                   1351
email                  1336
payment_contactless    1321
internet_access        1242
payment_debit_cards    1213
payment_credit_card    1206
payment_cash           1189
phone                  1085
website                 811
city                    314
postcode                277
housenumber             261
street                  253
brand                   224
wheelchair              145
opening_hours            98
name                      3
geometry                  0
latitude                  0
longitude                 0
dtype: int64


In [59]:
row_count = len(gdf_superstore)
print(row_count)
missing = pd.DataFrame({
    "missing_count": missing_count,
    "missing_pct": (missing_count / row_count * 100).round(1)
}).sort_values(by="missing_pct", ascending=False)

print(missing)

1358
                     missing_count  missing_pct
source                        1358        100.0
layer                         1355         99.8
type                          1351         99.5
email                         1336         98.4
payment_contactless           1321         97.3
internet_access               1242         91.5
payment_debit_cards           1213         89.3
payment_credit_card           1206         88.8
payment_cash                  1189         87.6
phone                         1085         79.9
website                        811         59.7
city                           314         23.1
postcode                       277         20.4
housenumber                    261         19.2
street                         253         18.6
brand                          224         16.5
wheelchair                     145         10.7
opening_hours                   98          7.2
name                             3          0.2
geometry                         0 

**✅ Recommendations**

- Drop or ignore high-missing fields

	•	Columns like source, layer, type, payment_*, and internet_access have too many missing values to be meaningful without external enrichment.

	•	You can either drop them or keep them for very limited exploratory use.

- Enrich moderate-missing fields

	•	Consider reverse geocoding to infer:

	•	street, housenumber, postcode, and city from latitude/longitude
	
	•	Use brand clustering or reference lists to fill common brand names.

- Fill remaining small gaps

	•	For opening_hours, wheelchair, and name, missing values can be flagged or marked as “unknown” for visualizations and charts.

- Keep spatial data clean

	•	Ensure geometry, latitude, and longitude are retained as the core fields for mapping and spatial joins.


## C) Distinct values per column

In [60]:
distinct = gdf_superstore.nunique().sort_values(ascending=False)
print(distinct)

longitude              1358
geometry               1358
latitude               1358
street                  635
housenumber             445
website                 444
name                    315
phone                   240
opening_hours           239
postcode                186
brand                    26
email                    22
type                      4
wheelchair                3
internet_access           3
payment_debit_cards       2
layer                     2
payment_credit_card       2
payment_cash              1
payment_contactless       1
city                      1
source                    0
dtype: int64


## D) Most common values in key columns

In [61]:
# Goal: Peek at distributions, not just counts.

# Example: top 10 brands
print("\nTop 10 brands:")
print(gdf_superstore["brand"].value_counts().head(10))


Top 10 brands:
brand
Edeka                    163
Lidl                     142
Netto Marken-Discount    126
Rewe                     121
Aldi Nord                 96
Penny                     71
EDEKA                     54
Denns BioMarkt            50
Bio Company               49
Netto                     44
Name: count, dtype: int64


In [62]:
print("\nTop 10 brands:")
print(gdf_superstore["street"].value_counts().head(10))


Top 10 brands:
street
Hauptstraße             19
Müllerstraße            14
Tempelhofer Damm        11
Frankfurter Allee       10
Wilmersdorfer Straße    10
Greifswalder Straße     10
Hermannstraße            9
Karl-Marx-Straße         9
Friedrichstraße          9
Landsberger Allee        8
Name: count, dtype: int64


In [63]:
print("\nTop 10 brands:")
print(gdf_superstore["opening_hours"].value_counts().head(10))


Top 10 brands:
opening_hours
Mo-Sa 07:00-22:00                    254
Mo-Sa 07:00-21:00                    247
Mo-Sa 08:00-21:00                    100
Mo-Sa 08:00-20:00                     88
Mo-Sa 07:00-20:00                     59
Mo-Sa 07:00-21:00; PH off             31
Mo-Sa 07:00-22:00; PH off             25
Mo-Sa 07:00-22:00; Su,PH off          23
Mo-Sa 08:00-22:00                     22
Mo-Fr 07:00-24:00; Sa 07:00-23:30     19
Name: count, dtype: int64


In [64]:
print("\nTop 10 brands:")
print(gdf_superstore["postcode"].value_counts().head(10))


Top 10 brands:
postcode
10827    15
10365    14
10117    13
10405    13
12683    12
10243    12
12524    12
12555    12
10245    12
10967    11
Name: count, dtype: int64


In [65]:
print("\nTop 10 brands:")
print(gdf_superstore["name"].value_counts().head(10))


Top 10 brands:
name
Lidl                     142
Aldi                     132
Netto Marken-Discount    126
REWE                     105
EDEKA                     82
PENNY                     51
Denns BioMarkt            50
Bio Company               49
Netto                     45
Kaufland                  35
Name: count, dtype: int64


## E) Geometry sanity checks

In [66]:
print(gdf_superstore.geometry.geom_type.value_counts())

Point    1358
Name: count, dtype: int64


In [67]:
print("Missing geometries:", gdf_superstore.geometry.isna().sum())

Missing geometries: 0


## F) Latitude/Longitude checks

In [68]:
# Goal: Verify lat/lon look realistic.
# Why? If values are way off, something went wrong in conversion.

print("Latitude range:", gdf_superstore["latitude"].min(), "to", gdf_superstore["latitude"].max())

print("Longitude range:", gdf_superstore["longitude"].min(), "to", gdf_superstore["longitude"].max())

Latitude range: 52.37918305338519 to 52.64074992624387
Longitude range: 13.124876595696755 to 13.7149761


## 1.3 Prepare the /sources Directory

1. Raw Data Files:

- banks_raw.geojson (includes geometry)
- banks_raw.csv (tabular only, no geometry)

2. README.md in /sources will contain:

- Data sources used.
- Planned transformation steps.

In [69]:
# Save as GeoJSON (keeps geometry) and CSV

raw_geojson_path = "../sources/supermarkets_raw.geojson"
raw_csv_path = "../sources/supermarkets_raw.csv"


gdf_superstore.to_file(raw_geojson_path, driver="GeoJSON")
gdf_superstore.drop(columns="geometry").to_csv(raw_csv_path, index=False)

print(f"Raw data saved to: {raw_geojson_path} and {raw_csv_path}")

Raw data saved to: ../sources/supermarkets_raw.geojson and ../sources/supermarkets_raw.csv


# 1.4 Review

- All 22 target columns defined.
- Data sources identified and documented.
- Schema draft created.
- Data fetched and stored in /sources.
- Data cleaning & enrichment plan in place.

# 🛠 Step 2: Data Transformation

- Table Finalization & Test Insert

Task for Data Transformation Part:


- Write transformation logic using Python/SQL.

- Ensure data is clean, normalized, and matches the model defined in Step 1.

- Do not insert into the DB yet — this step is focused purely on transformation.

- Place scripts in the /scripts folder.

- Test transformation outputs locally.

- Submit the second PRs

**🛠️ Data Transformation Goals – Summary**

**1. ✅ Match Target Schema**

- Use correct column names (snake_case)
- Ensure all required fields are present (e.g., name, latitude, longitude, district, etc.)

**2. 🧹 Clean & Normalize**

- Strip spaces, fix casing (e.g., rewe → REWE)
- Standardize formats (e.g., opening hours, addresses)

**3. ❓ Handle Missing Data**

- Fill missing address info using reverse geolocation
- Assign district and neighborhood via spatial join
- Drop rows only if critical fields are missing

**4. 🧮 Convert & Validate Data Types**

- Latitude/longitude → float
- Postal code → str
- Capacity → int or float
- Geometry → EPSG:4326 (if using GeoDataFrame)

**5. ➕ Add Derived Columns**

- full_address from reverse geolocation
- osm_id, district, neighborhood, etc.

**6. 🔍 Run Quality Checks**

- df.info(), df.isna().sum(), df.nunique()
- Visual check with df.head()

**7. 💾 Save Transformed Data**

- Export clean output as .csv and .geojson
- Store script in /scripts/ folder

**8. 🧪 Test Locally Before Inserting to DB**


## A) Standardize column names and types

In [70]:
gdf_superstore.head(3)

name            street housenumber  \
element id                                                              
node    58489979  Netto Marken-Discount  Alte Jakobstraße          83   
        79418658                   Ledo      Quäkerstraße           2   
        79422426              kiezmarkt               NaN         NaN   

                 postcode    city                      opening_hours  \
element id                                                             
node    58489979    10179  Berlin          Mo-Sa 07:00-24:00; Su off   
        79418658    13403  Berlin                  Mo-Sa 09:00-20:00   
        79422426      NaN     NaN  Mo-Fr 08:00-20:00, Sa 08:00-19:00   

                                  brand type                   geometry  \
element id                                                                
node    58489979  Netto Marken-Discount  NaN  POINT (13.40737 52.50982)   
        79418658                    NaN  NaN  POINT (13.30855 52.57242)   
        79422426                Nahkauf  NaN  POINT (13.31248 52.57168)   

                 payment_credit_card  ... payment_contactless wheelchair  \
element id                            ...                                  
node    58489979                 NaN  ...                 NaN         no   
        79418658                 NaN  ...                 NaN        yes   
        79422426                 NaN  ...                 NaN        yes   

                 internet_access layer phone email  \
element id                                           
node    58489979              no   NaN   NaN   NaN   
        79418658             NaN   NaN   NaN   NaN   
        79422426             NaN   NaN   NaN   NaN   

                                                            website  \
element id                                                            
node    58489979  https://www.netto-online.de/filialen/Berlin-Mi...   
        79418658                    https://www.ledo-supermarkt.de/   
        79422426                                                NaN   

                   latitude  longitude  source  
element id                                      
node    58489979  52.509819  13.407373    None  
        79418658  52.572417  13.308547    None  
        79422426  52.571676  13.312476    None  

[3 rows x 22 columns]

In [35]:
gdf_superstore.dtypes

name                     object
street                   object
housenumber              object
postcode                 object
city                     object
opening_hours            object
brand                    object
type                     object
geometry               geometry
payment_credit_card      object
payment_debit_cards      object
payment_cash             object
payment_contactless      object
wheelchair               object
internet_access          object
layer                    object
phone                    object
email                    object
website                  object
latitude                float64
longitude               float64
source                   object
dtype: object

In [74]:
# Standardize column names
gdf_superstore = gdf_superstore.rename(columns={
    "name": "store_name"
    })
gdf_superstore.columns = gdf_superstore.columns.str.lower().str.strip().str.replace(" ", "_").str.replace("-", "_")

# Convert certain columns to correct type

# Convert columns to specific types
df_transformed = gdf_superstore.copy()

# String columns (ensure consistent text type)
string_columns = [
    "store_name", "street", "housenumber", "postcode", "city", "opening_hours",
    "brand", "type", "payment_credit_card", "payment_debit_cards", "payment_cash",
    "payment_contactless", "wheelchair", "internet_access", "layer",
    "phone", "email", "website", "source"
]
df_transformed[string_columns] = df_transformed[string_columns].apply(lambda col: col.astype("string").str.strip().str.lower()) # ensure text type

# Latitude & longitude to float
df_transformed["latitude"] = pd.to_numeric(df_transformed["latitude"], errors="coerce")
df_transformed["longitude"] = pd.to_numeric(df_transformed["longitude"], errors="coerce")

# Preview the updated dtypes
print(df_transformed.dtypes)

store_name             string[python]
street                 string[python]
housenumber            string[python]
postcode               string[python]
city                   string[python]
opening_hours          string[python]
brand                  string[python]
type                   string[python]
geometry                     geometry
payment_credit_card    string[python]
payment_debit_cards    string[python]
payment_cash           string[python]
payment_contactless    string[python]
wheelchair             string[python]
internet_access        string[python]
layer                  string[python]
phone                  string[python]
email                  string[python]
website                string[python]
latitude                      float64
longitude                     float64
source                 string[python]
dtype: object


In [78]:
# Display all columns
pd.set_option('display.max_columns', None)

# Display the first 3 rows of the dataframe
df_transformed.head(3)

store_name            street housenumber  \
element id                                                              
node    58489979  netto marken-discount  alte jakobstraße          83   
        79418658                   ledo      quäkerstraße           2   
        79422426              kiezmarkt              <NA>        <NA>   

                 postcode    city                      opening_hours  \
element id                                                             
node    58489979    10179  berlin          mo-sa 07:00-24:00; su off   
        79418658    13403  berlin                  mo-sa 09:00-20:00   
        79422426     <NA>    <NA>  mo-fr 08:00-20:00, sa 08:00-19:00   

                                  brand  type                   geometry  \
element id                                                                 
node    58489979  netto marken-discount  <NA>  POINT (13.40737 52.50982)   
        79418658                   <NA>  <NA>  POINT (13.30855 52.57242)   
        79422426                nahkauf  <NA>  POINT (13.31248 52.57168)   

                 payment_credit_card payment_debit_cards payment_cash  \
element id                                                              
node    58489979                <NA>                <NA>         <NA>   
        79418658                <NA>                <NA>         <NA>   
        79422426                <NA>                <NA>         <NA>   

                 payment_contactless wheelchair internet_access layer phone  \
element id                                                                    
node    58489979                <NA>         no              no  <NA>  <NA>   
        79418658                <NA>        yes            <NA>  <NA>  <NA>   
        79422426                <NA>        yes            <NA>  <NA>  <NA>   

                 email                                            website  \
element id                                                                  
node    58489979  <NA>  https://www.netto-online.de/filialen/berlin-mi...   
        79418658  <NA>                    https://www.ledo-supermarkt.de/   
        79422426  <NA>                                               <NA>   

                   latitude  longitude source  
element id                                     
node    58489979  52.509819  13.407373   <NA>  
        79418658  52.572417  13.308547   <NA>  
        79422426  52.571676  13.312476   <NA>

## B) Fixing Missing Values

In [ ]:
# missing value analysis
# Check for missing values in percentage
(df_transformed.isna().mean() * 100).sort_values(ascending=False)

source                 100.000000
layer                   99.779087
type                    99.484536
email                   98.379971
payment_contactless     97.275405
internet_access         91.458027
payment_debit_cards     89.322533
payment_credit_card     88.807069
payment_cash            87.555228
phone                   79.896907
website                 59.720177
city                    23.122239
postcode                20.397644
housenumber             19.219440
street                  18.630339
brand                   16.494845
wheelchair              10.677467
opening_hours            7.216495
store_name               0.220913
geometry                 0.000000
latitude                 0.000000
longitude                0.000000
dtype: float64

In [82]:
# 1. Fill constant values
df_transformed['source'] = 'OSM'

# 2. Drop unhelpful column
df_transformed.drop(columns=['layer'], inplace=True)

# 3. Fill common placeholders
fill_with_unknown = [
    'type', 'payment_credit_card', 'payment_debit_cards',
    'payment_cash', 'payment_contactless', 'internet_access',
    'wheelchair', 'opening_hours'
]

df_transformed[fill_with_unknown] = df_transformed[fill_with_unknown].fillna('unknown')

# 4. Fill optional contact info
df_transformed[['phone', 'email', 'website']] = df_transformed[['phone', 'email', 'website']].fillna('not_provided')

# 5. Address components: try reverse geocoding
address_cols = ['street', 'housenumber', 'postcode', 'city']
#df[address_cols] = df[address_cols].fillna('unknown')

# 6. Brand
df_transformed['brand'] = df_transformed['brand'].fillna('independent')

# 7. Drop rows where store name is missing (critical field)
df_transformed.dropna(subset=['store_name'], inplace=True)

In [83]:
df_transformed.shape

(1355, 21)

In [84]:
# Reverse Geocoding Missing Address Fields Using Nominatim

import requests
import time
import pandas as pd
import logging

def reverse_geocode(lat, lon):
    """Use Nominatim to get address details for a given coordinate"""
    url = "https://nominatim.openstreetmap.org/reverse"
    params = {
        "lat": lat,
        "lon": lon,
        "format": "json",
        "addressdetails": 1
    }
    headers = {
        "User-Agent": "berlin-superstore-cleaner/1.0"
    }
    try:
        response = requests.get(url, params=params, headers=headers, timeout=10)
        response.raise_for_status()
        data = response.json().get("address", {})
        return {
            "street": data.get("road"),
            "housenumber": data.get("house_number"),
            "postcode": data.get("postcode"),
            "city": data.get("city") or data.get("town") or data.get("village")
        }
    except requests.exceptions.RequestException as e:
        logging.warning(f"Reverse geocode failed for ({lat}, {lon}): {e}")
        return {
            "street": None,
            "housenumber": None,
            "postcode": None,
            "city": None
        }

In [85]:
# apply to rows with missing address components

# Loop through rows with any missing address field
for i, row in df_transformed[
    df_transformed[["street", "housenumber", "postcode", "city"]].isnull().any(axis=1)
].iterrows():
    
    lat, lon = row["latitude"], row["longitude"]
    print(f"⏳ Fetching address for index {i}...")
    
    # Get address
    addr = reverse_geocode(lat, lon)
    
    # Fill in only if missing
    for field in ["street", "housenumber", "postcode", "city"]:
        if pd.isna(df_transformed.at[i, field]) and addr[field] is not None:
            df_transformed.at[i, field] = addr[field]
    
    time.sleep(1)  # Throttle per Nominatim's usage policy

⏳ Fetching address for index ('node', 79422426)...
⏳ Fetching address for index ('node', 151015050)...
⏳ Fetching address for index ('node', 175685784)...
⏳ Fetching address for index ('node', 249675570)...
⏳ Fetching address for index ('node', 251595302)...
⏳ Fetching address for index ('node', 253256553)...
⏳ Fetching address for index ('node', 253616655)...
⏳ Fetching address for index ('node', 254862675)...
⏳ Fetching address for index ('node', 268844516)...
⏳ Fetching address for index ('node', 268845115)...
⏳ Fetching address for index ('node', 270793632)...
⏳ Fetching address for index ('node', 273759099)...
⏳ Fetching address for index ('node', 279224131)...
⏳ Fetching address for index ('node', 283190905)...
⏳ Fetching address for index ('node', 293130845)...
⏳ Fetching address for index ('node', 293130996)...
⏳ Fetching address for index ('node', 293137288)...
⏳ Fetching address for index ('node', 301787391)...
⏳ Fetching address for index ('node', 313144179)...
⏳ Fetching ad

In [86]:
df_transformed.isnull().mean() * 100

store_name              0.000000
street                  0.295203
housenumber            16.605166
postcode                0.000000
city                    0.000000
opening_hours           0.000000
brand                   0.000000
type                    0.000000
geometry                0.000000
payment_credit_card     0.000000
payment_debit_cards     0.000000
payment_cash            0.000000
payment_contactless     0.000000
wheelchair              0.000000
internet_access         0.000000
phone                   0.000000
email                   0.000000
website                 0.000000
latitude                0.000000
longitude               0.000000
source                  0.000000
dtype: float64

In [87]:
# Fill any remaining NAs with fallback
df_transformed[["street", "housenumber", "postcode", "city"]] = df_transformed[
    ["street", "housenumber", "postcode", "city"]
].fillna("unknown")

In [88]:
df_transformed.isnull().mean() * 100

store_name             0.0
street                 0.0
housenumber            0.0
postcode               0.0
city                   0.0
opening_hours          0.0
brand                  0.0
type                   0.0
geometry               0.0
payment_credit_card    0.0
payment_debit_cards    0.0
payment_cash           0.0
payment_contactless    0.0
wheelchair             0.0
internet_access        0.0
phone                  0.0
email                  0.0
website                0.0
latitude               0.0
longitude              0.0
source                 0.0
dtype: float64

In [89]:
df_transformed.sample(10, random_state=42)

store_name                street  \
element id                                                             
node    292092618                        edeka    franz-jacob-straße   
way     64400029                         penny     mariendorfer damm   
node    2866172985                       netto      ribnitzer straße   
way     79308490    penny oranienburger straße  oranienburger straße   
        26819746                          lidl      gutschmidtstraße   
node    7734882729                        lidl           hauptstraße   
        747081738                        penny    brunsbütteler damm   
way     28464064                          lidl          säntisstraße   
node    5943512628       netto marken-discount         gotlandstraße   
        4868088443                       netto        mehrower allee   

                   housenumber postcode    city                 opening_hours  \
element id                                                                      
node    292092618            4    10369  berlin             mo-sa 07:00-22:00   
way     64400029       153-157    12107  berlin             mo-sa 07:00-22:00   
node    2866172985          24    13051  berlin             mo-sa 08:00-21:00   
way     79308490           116    13437  berlin             mo-sa 07:00-22:00   
        26819746           101    12359  Berlin             mo-sa 07:00-21:00   
node    7734882729          65    12159  berlin             mo-sa 07:00-22:00   
        747081738          196    13581  berlin  mo-sa 07:00-22:00; su,ph off   
way     28464064            49    12277  berlin             mo-sa 07:00-21:00   
node    5943512628           6    10439  berlin             mo-sa 07:00-22:00   
        4868088443          20    12687  berlin             mo-sa 08:00-20:00   

                                    brand     type                   geometry  \
element id                                                                      
node    292092618                   edeka  unknown  POINT (13.46573 52.52543)   
way     64400029                    penny  unknown  POINT (13.38888 52.43793)   
node    2866172985                  netto  unknown  POINT (13.49599 52.57027)   
way     79308490                    penny  unknown   POINT (13.3312 52.60217)   
        26819746                     lidl  unknown  POINT (13.43672 52.43945)   
node    7734882729                   lidl  unknown   POINT (13.34162 52.4771)   
        747081738                   penny  unknown    POINT (13.16773 52.534)   
way     28464064                     lidl  unknown  POINT (13.39074 52.41917)   
node    5943512628  netto marken-discount  unknown  POINT (13.41135 52.55476)   
        4868088443                  netto  unknown  POINT (13.55883 52.55566)   

                   payment_credit_card payment_debit_cards payment_cash  \
element id                                                                
node    292092618              unknown             unknown      unknown   
way     64400029               unknown             unknown      unknown   
node    2866172985             unknown             unknown      unknown   
way     79308490               unknown             unknown      unknown   
        26819746               unknown             unknown      unknown   
node    7734882729             unknown             unknown      unknown   
        747081738              unknown             unknown      unknown   
way     28464064               unknown             unknown      unknown   
node    5943512628                 yes                 yes          yes   
        4868088443             unknown             unknown      unknown   

                   payment_contactless wheelchair internet_access  \
element id                                                          
node    292092618              unknown        yes         unknown   
way     64400029               unknown        yes         unknown   
node    2866172985             unknown        ye

## C) Add district, district_id , neighbor, neighborhood_id to the data frame

**Approach: Use Pre-downloaded GeoJSON Boundary File**

I have already a GeoJSON file like lor_ortsteile.geojson, which contains:
- district (Bezirk)
- neighborhood (Ortsteil)
- neighborhood_id (spatial ID)

In [90]:
import geopandas as gpd

# Load the GeoJSON file
# neighborhoods = gpd.read_file("lor_ortsteile.geojson").to_crs("EPSG:4326")


# Load GeoJSON using relative path
geojson_path = "../sources/lor_ortsteile.geojson"
neighborhoods = gpd.read_file(geojson_path).to_crs("EPSG:4326")

# Optional renaming for consistency
neighborhoods = neighborhoods.rename(columns={
    "BEZIRK": "district",
    "OTEIL": "neighborhood",
    "spatial_name": "neighborhood_id"
})

# Preview
neighborhoods.head()

gml_id neighborhood_id spatial_alias spatial_type  neighborhood  \
0  re_ortsteil.0101            0101         Mitte      Polygon         Mitte   
1  re_ortsteil.0102            0102        Moabit      Polygon        Moabit   
2  re_ortsteil.0103            0103  Hansaviertel      Polygon  Hansaviertel   
3  re_ortsteil.0104            0104    Tiergarten      Polygon    Tiergarten   
4  re_ortsteil.0105            0105       Wedding      Polygon       Wedding   

  district  FLAECHE_HA                                           geometry  
0    Mitte   1063.8748  POLYGON ((13.41649 52.52696, 13.41635 52.52702...  
1    Mitte    768.7909  POLYGON ((13.33884 52.51974, 13.33884 52.51974...  
2    Mitte     52.5337  POLYGON ((13.34322 52.51557, 13.34323 52.51557...  
3    Mitte    516.0672  POLYGON ((13.36879 52.49878, 13.36891 52.49877...  
4    Mitte    919.9112  POLYGON ((13.34656 52.53879, 13.34664 52.53878...

In [128]:
# spatially join this with store using gpd.sjoin()
df_clean=df_transformed.copy()
df_with_districts = gpd.sjoin(
    df_clean, 
    neighborhoods[["district", "neighborhood", "neighborhood_id", "geometry"]],
    how="left",
    predicate="within"
)

# Drop unnecessary columns
df_final = df_with_districts.drop(columns=["geometry", "index_right"])

In [129]:
df_final.head(3)

store_name                  street housenumber  \
element id                                                                    
node    58489979  netto marken-discount        alte jakobstraße          83   
        79418658                   ledo            quäkerstraße           2   
        79422426              kiezmarkt  Auguste-Viktoria-Allee     unknown   

                 postcode    city                      opening_hours  \
element id                                                             
node    58489979    10179  berlin          mo-sa 07:00-24:00; su off   
        79418658    13403  berlin                  mo-sa 09:00-20:00   
        79422426    13403  Berlin  mo-fr 08:00-20:00, sa 08:00-19:00   

                                  brand     type payment_credit_card  \
element id                                                             
node    58489979  netto marken-discount  unknown             unknown   
        79418658            independent  unknown             unknown   
        79422426                nahkauf  unknown             unknown   

                 payment_debit_cards payment_cash payment_contactless  \
element id                                                              
node    58489979             unknown      unknown             unknown   
        79418658             unknown      unknown             unknown   
        79422426             unknown      unknown             unknown   

                 wheelchair internet_access         phone         email  \
element id                                                                
node    58489979         no              no  not_provided  not_provided   
        79418658        yes         unknown  not_provided  not_provided   
        79422426        yes         unknown  not_provided  not_provided   

                                                            website  \
element id                                                            
node    58489979  https://www.netto-online.de/filialen/berlin-mi...   
        79418658                    https://www.ledo-supermarkt.de/   
        79422426                                       not_provided   

                   latitude  longitude source       district   neighborhood  \
element id                                                                    
node    58489979  52.509819  13.407373    OSM          Mitte          Mitte   
        79418658  52.572417  13.308547    OSM  Reinickendorf  Reinickendorf   
        79422426  52.571676  13.312476    OSM  Reinickendorf  Reinickendorf   

                 neighborhood_id  
element id                        
node    58489979            0101  
        79418658            1201  
        79422426            1201

In [130]:
# Generating district ids
# https://www.regionalstatistik.de

# District mapping (official codes as strings)
district_mapping = {
    'Mitte': '11001001',
    'Friedrichshain-Kreuzberg': '11002002',
    'Pankow': '11003003',
    'Charlottenburg-Wilmersdorf': '11004004',
    'Spandau': '11005005',
    'Steglitz-Zehlendorf': '11006006',
    'Tempelhof-Schöneberg': '11007007',
    'Neukölln': '11008008',
    'Treptow-Köpenick': '11009009',
    'Marzahn-Hellersdorf': '11010010',
    'Lichtenberg': '11011011',
    'Reinickendorf': '11012012'
}

# Apply mapping to create district_id column (string)
df_final['district_id'] = df_final['district'].map(district_mapping).astype(str)

# (Optional) Check if some districts were not mapped
unmapped = df_final[~df_final['district'].isin(district_mapping.keys())]['district'].unique()
if len(unmapped) > 0:
    print("⚠️ Unmapped districts found:", unmapped)

In [131]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 1355 entries, ('node', np.int64(58489979)) to ('way', np.int64(1429504461))
Data columns (total 24 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   store_name           1355 non-null   string 
 1   street               1355 non-null   string 
 2   housenumber          1355 non-null   string 
 3   postcode             1355 non-null   string 
 4   city                 1355 non-null   string 
 5   opening_hours        1355 non-null   string 
 6   brand                1355 non-null   string 
 7   type                 1355 non-null   string 
 8   payment_credit_card  1355 non-null   string 
 9   payment_debit_cards  1355 non-null   string 
 10  payment_cash         1355 non-null   string 
 11  payment_contactless  1355 non-null   string 
 12  wheelchair           1355 non-null   string 
 13  internet_access      1355 non-null   string 
 14  phone                1355 non-null  

In [132]:
# change certain columns to string type
cols_to_string = ["source", "district", "neighborhood", "neighborhood_id", "district_id"]
df_final[cols_to_string] = df_final[cols_to_string].astype("string")
# normalize the text to lowercase and strip whitespace
df_final[cols_to_string] = df_final[cols_to_string].apply(lambda col: col.str.strip().str.lower())

In [133]:
df_final.sample(5, random_state=42)

store_name                street  \
element id                                                             
node    292092618                        edeka    franz-jacob-straße   
way     64400029                         penny     mariendorfer damm   
node    2866172985                       netto      ribnitzer straße   
way     79308490    penny oranienburger straße  oranienburger straße   
        26819746                          lidl      gutschmidtstraße   

                   housenumber postcode    city      opening_hours  brand  \
element id                                                                  
node    292092618            4    10369  berlin  mo-sa 07:00-22:00  edeka   
way     64400029       153-157    12107  berlin  mo-sa 07:00-22:00  penny   
node    2866172985          24    13051  berlin  mo-sa 08:00-21:00  netto   
way     79308490           116    13437  berlin  mo-sa 07:00-22:00  penny   
        26819746           101    12359  Berlin  mo-sa 07:00-21:00   lidl   

                       type payment_credit_card payment_debit_cards  \
element id                                                            
node    292092618   unknown             unknown             unknown   
way     64400029    unknown             unknown             unknown   
node    2866172985  unknown             unknown             unknown   
way     79308490    unknown             unknown             unknown   
        26819746    unknown             unknown             unknown   

                   payment_cash payment_contactless wheelchair  \
element id                                                       
node    292092618       unknown             unknown        yes   
way     64400029        unknown             unknown        yes   
node    2866172985      unknown             unknown        yes   
way     79308490        unknown             unknown        yes   
        26819746        unknown             unknown    limited   

                   internet_access               phone         email  \
element id                                                             
node    292092618          unknown        not_provided  not_provided   
way     64400029           unknown  +49 221 201 999 59  not_provided   
node    2866172985         unknown        not_provided  not_provided   
way     79308490           unknown        not_provided  not_provided   
        26819746           unknown        not_provided  not_provided   

                                                              website  \
element id                                                              
node    292092618                                        not_provided   
way     64400029                                https://www.penny.de/   
node    2866172985  https://netto.de/geschaefte/berlin/ribnitzer-s...   
way     79308490                                https://www.penny.de/   
        26819746                                         not_provided   

                     latitude  longitude source              district  \
element id                                                              
node    292092618   52.525430  13.465731    osm           lichtenberg   
way     64400029    52.437925  13.388883    osm  tempelhof-schöneberg   
node    2866172985  52.570273  13.495987    osm           lichtenberg   
way     79308490    52.602177  13.331190    osm         reinickendorf   
        26819746    52.439440  13.436725    osm              neukölln   

                            neighborhood neighborhood_id district_id  
element id                                                            
node    292092618              fennpfuhl            1111    11011011  
way     64400029              mariendorf            0704    11007007  
node    2866172985  neu-hohenschönhausen            1109    11011011  
way     79308490                wittenau            1209    11012012  
        26819746                   britz            0802    11008008

## D) Reset index, drop columns "element" and  rename "id" to "store_id"

In [134]:
# Reset index to move 'element' and 'id' into columns
df_final = df_final.reset_index()
df_final.head(2)

element        id             store_name            street housenumber  \
0    node  58489979  netto marken-discount  alte jakobstraße          83   
1    node  79418658                   ledo      quäkerstraße           2   

  postcode    city              opening_hours                  brand     type  \
0    10179  berlin  mo-sa 07:00-24:00; su off  netto marken-discount  unknown   
1    13403  berlin          mo-sa 09:00-20:00            independent  unknown   

  payment_credit_card payment_debit_cards payment_cash payment_contactless  \
0             unknown             unknown      unknown             unknown   
1             unknown             unknown      unknown             unknown   

  wheelchair internet_access         phone         email  \
0         no              no  not_provided  not_provided   
1        yes         unknown  not_provided  not_provided   

                                             website   latitude  longitude  \
0  https://www.netto-online.de/filialen/berlin-mi...  52.509819  13.407373   
1                    https://www.ledo-supermarkt.de/  52.572417  13.308547   

  source       district   neighborhood neighborhood_id district_id  
0    osm          mitte          mitte            0101    11001001  
1    osm  reinickendorf  reinickendorf            1201    11012012

In [135]:
# Drop the 'element' column and rename 'id' to 'store_id'
df_final = df_final.drop(columns=["element"]).rename(columns={"id": "store_id"})
df_final["store_id"] = df_final["store_id"].astype("string")

In [136]:
df_final.sample(5, random_state=42)

store_id                  store_name                street  \
49     292092618                       edeka    franz-jacob-straße   
1096    64400029                       penny     mariendorfer damm   
494   2866172985                       netto      ribnitzer straße   
1110    79308490  penny oranienburger straße  oranienburger straße   
922     26819746                        lidl      gutschmidtstraße   

     housenumber postcode    city      opening_hours  brand     type  \
49             4    10369  berlin  mo-sa 07:00-22:00  edeka  unknown   
1096     153-157    12107  berlin  mo-sa 07:00-22:00  penny  unknown   
494           24    13051  berlin  mo-sa 08:00-21:00  netto  unknown   
1110         116    13437  berlin  mo-sa 07:00-22:00  penny  unknown   
922          101    12359  Berlin  mo-sa 07:00-21:00   lidl  unknown   

     payment_credit_card payment_debit_cards payment_cash payment_contactless  \
49               unknown             unknown      unknown             unknown   
1096             unknown             unknown      unknown             unknown   
494              unknown             unknown      unknown             unknown   
1110             unknown             unknown      unknown             unknown   
922              unknown             unknown      unknown             unknown   

     wheelchair internet_access               phone         email  \
49          yes         unknown        not_provided  not_provided   
1096        yes         unknown  +49 221 201 999 59  not_provided   
494         yes         unknown        not_provided  not_provided   
1110        yes         unknown        not_provided  not_provided   
922     limited         unknown        not_provided  not_provided   

                                                website   latitude  longitude  \
49                                         not_provided  52.525430  13.465731   
1096                              https://www.penny.de/  52.437925  13.388883   
494   https://netto.de/geschaefte/berlin/ribnitzer-s...  52.570273  13.495987   
1110                              https://www.penny.de/  52.602177  13.331190   
922                                        not_provided  52.439440  13.436725   

     source              district          neighborhood neighborhood_id  \
49      osm           lichtenberg             fennpfuhl            1111   
1096    osm  tempelhof-schöneberg            mariendorf            0704   
494     osm           lichtenberg  neu-hohenschönhausen            1109   
1110    osm         reinickendorf              wittenau            1209   
922     osm              neukölln                 britz            0802   

     district_id  
49      11011011  
1096    11007007  
494     11011011  
1110    11012012  
922     11008008

In [137]:
df_final.head(2)

store_id             store_name            street housenumber postcode  \
0  58489979  netto marken-discount  alte jakobstraße          83    10179   
1  79418658                   ledo      quäkerstraße           2    13403   

     city              opening_hours                  brand     type  \
0  berlin  mo-sa 07:00-24:00; su off  netto marken-discount  unknown   
1  berlin          mo-sa 09:00-20:00            independent  unknown   

  payment_credit_card payment_debit_cards payment_cash payment_contactless  \
0             unknown             unknown      unknown             unknown   
1             unknown             unknown      unknown             unknown   

  wheelchair internet_access         phone         email  \
0         no              no  not_provided  not_provided   
1        yes         unknown  not_provided  not_provided   

                                             website   latitude  longitude  \
0  https://www.netto-online.de/filialen/berlin-mi...  52.509819  13.407373   
1                    https://www.ledo-supermarkt.de/  52.572417  13.308547   

  source       district   neighborhood neighborhood_id district_id  
0    osm          mitte          mitte            0101    11001001  
1    osm  reinickendorf  reinickendorf            1201    11012012

In [138]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1355 entries, 0 to 1354
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   store_id             1355 non-null   string 
 1   store_name           1355 non-null   string 
 2   street               1355 non-null   string 
 3   housenumber          1355 non-null   string 
 4   postcode             1355 non-null   string 
 5   city                 1355 non-null   string 
 6   opening_hours        1355 non-null   string 
 7   brand                1355 non-null   string 
 8   type                 1355 non-null   string 
 9   payment_credit_card  1355 non-null   string 
 10  payment_debit_cards  1355 non-null   string 
 11  payment_cash         1355 non-null   string 
 12  payment_contactless  1355 non-null   string 
 13  wheelchair           1355 non-null   string 
 14  internet_access      1355 non-null   string 
 15  phone                1355 non-null   s

In [144]:
# (Optional) Save enriched dataset for later use
df_final.to_csv("../sources/final_supermarkets_with_district.csv", index=False)

## E) Final Summary of Cleaned and Transformed Data

In [141]:
print("✅ Dataset after Steps A - D cleaning and transforming\n")

# Shape of dataframe
print(f"Number of rows: {df_final.shape[0]}")
print(f"Number of columns: {df_final.shape[1]}")

# Column list
print("\nRemaining columns:")
print(df_final.columns.tolist())

# Missing values check
missing = df_final.isnull().sum()
print("\nMissing values after cleaning and transforming :")
print(missing)

✅ Dataset after Steps A - D cleaning and transforming

Number of rows: 1355
Number of columns: 25

Remaining columns:
['store_id', 'store_name', 'street', 'housenumber', 'postcode', 'city', 'opening_hours', 'brand', 'type', 'payment_credit_card', 'payment_debit_cards', 'payment_cash', 'payment_contactless', 'wheelchair', 'internet_access', 'phone', 'email', 'website', 'latitude', 'longitude', 'source', 'district', 'neighborhood', 'neighborhood_id', 'district_id']

Missing values after cleaning and transforming :
store_id               0
store_name             0
street                 0
housenumber            0
postcode               0
city                   0
opening_hours          0
brand                  0
type                   0
payment_credit_card    0
payment_debit_cards    0
payment_cash           0
payment_contactless    0
wheelchair             0
internet_access        0
phone                  0
email                  0
website                0
latitude               0
longitu

In [142]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1355 entries, 0 to 1354
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   store_id             1355 non-null   string 
 1   store_name           1355 non-null   string 
 2   street               1355 non-null   string 
 3   housenumber          1355 non-null   string 
 4   postcode             1355 non-null   string 
 5   city                 1355 non-null   string 
 6   opening_hours        1355 non-null   string 
 7   brand                1355 non-null   string 
 8   type                 1355 non-null   string 
 9   payment_credit_card  1355 non-null   string 
 10  payment_debit_cards  1355 non-null   string 
 11  payment_cash         1355 non-null   string 
 12  payment_contactless  1355 non-null   string 
 13  wheelchair           1355 non-null   string 
 14  internet_access      1355 non-null   string 
 15  phone                1355 non-null   s

In [143]:
df_final.head(3)

store_id             store_name                  street housenumber  \
0  58489979  netto marken-discount        alte jakobstraße          83   
1  79418658                   ledo            quäkerstraße           2   
2  79422426              kiezmarkt  Auguste-Viktoria-Allee     unknown   

  postcode    city                      opening_hours                  brand  \
0    10179  berlin          mo-sa 07:00-24:00; su off  netto marken-discount   
1    13403  berlin                  mo-sa 09:00-20:00            independent   
2    13403  Berlin  mo-fr 08:00-20:00, sa 08:00-19:00                nahkauf   

      type payment_credit_card payment_debit_cards payment_cash  \
0  unknown             unknown             unknown      unknown   
1  unknown             unknown             unknown      unknown   
2  unknown             unknown             unknown      unknown   

  payment_contactless wheelchair internet_access         phone         email  \
0             unknown         no              no  not_provided  not_provided   
1             unknown        yes         unknown  not_provided  not_provided   
2             unknown        yes         unknown  not_provided  not_provided   

                                             website   latitude  longitude  \
0  https://www.netto-online.de/filialen/berlin-mi...  52.509819  13.407373   
1                    https://www.ledo-supermarkt.de/  52.572417  13.308547   
2                                       not_provided  52.571676  13.312476   

  source       district   neighborhood neighborhood_id district_id  
0    osm          mitte          mitte            0101    11001001  
1    osm  reinickendorf  reinickendorf            1201    11012012  
2    osm  reinickendorf  reinickendorf            1201    11012012

**🛠️ Data Transformation Summary**

**1. 📦 Source Data:**

•	Extracted supermarket data from OpenStreetMap (OSM) using the osmnx package.
•	Included metadata such as name, address, opening hours, brand, payment types, wheelchair access, and location (lat/lon).

**2. 🧹 Cleaning & Normalization:**

•	Renamed columns for consistency (snake_case, lowercased, trimmed).
•	Converted appropriate fields to correct data types (string, float64).
•	Handled missing values (reverse geocoding for address fields).
•	Harmonized text fields (e.g., lowercased city names like “berlin”).
•	Dropped irrelevant or redundant columns (element, unused geometry after extraction).

**3. 🌍 Geo-enrichment:**

•	Assigned each supermarket to a district and neighborhood using spatial joins (lor_ortsteile.geojson).
•	Ensured latitude/longitude were extracted from geometry.

**4. 🧪 Validation & Export:**

•	Verified schema via .info(), column types, missing data checks.
•	Exported cleaned data to:
•	CSV: ../sources/raw_supermarkets.csv
•	(Geo)JSON if needed.

# Step 3: Populate dataset into neondb Tester Database

**🛠️ 1. Connect to the Neon PostgreSQL DB (e.g. with SQLAlchemy)**

In [9]:
import pandas as pd
from sqlalchemy import create_engine, text
import warnings

warnings.filterwarnings("ignore")

# Step 1: Neon DB connection string
DATABASE_URL = (
    "postgresql+psycopg2://neondb_owner:a9Am7Yy5r9_T7h4OF2GN@ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech:5432/neondb?sslmode=require"
)

# Step 2: Create engine
engine = create_engine(DATABASE_URL)

# Step 3: Drop old table if needed (be careful)
drop_sql = "DROP TABLE IF EXISTS test_berlin_data.supermarkets_in_berlin;"
create_sql = """
CREATE TABLE test_berlin_data.supermarkets_in_berlin (
    store_id TEXT PRIMARY KEY,
    store_name TEXT,
    street TEXT,
    housenumber TEXT,
    postcode TEXT,
    city TEXT,
    opening_hours TEXT,
    brand TEXT,
    type TEXT,
    payment_credit_card TEXT,
    payment_debit_cards TEXT,
    payment_cash TEXT,
    payment_contactless TEXT,
    wheelchair TEXT,
    internet_access TEXT,
    phone TEXT,
    email TEXT,
    website TEXT,
    latitude DOUBLE PRECISION,
    longitude DOUBLE PRECISION,
    source TEXT,
    district TEXT,
    neighborhood TEXT,
    neighborhood_id TEXT,
    district_id TEXT
);
"""

with engine.connect() as conn:
    conn.execute(text(drop_sql))
    conn.execute(text(create_sql))
    print("✅ Table recreated successfully with PK.")

# Step 4: Load DataFrame
df_final = pd.read_csv("../sources/final_supermarkets_with_district.csv")

# Step 5: Append data
df_final.to_sql(
    name="supermarkets_in_berlin",
    con=engine,
    schema="test_berlin_data",
    if_exists="append",   # now safe to append
    index=False
)
print(f"✅ Inserted {len(df_final)} rows into DB.")


✅ Table recreated successfully with PK.
✅ Inserted 1355 rows into DB.


**📥 2. Insert Clean DataFrame (df_final) into supermarkets_in_berlin Table**

In [10]:
# Assuming df_final is your cleaned and transformed DataFrame
df_final= pd.read_csv("../sources/final_supermarkets_with_district.csv")

df_final.to_sql(
    name="supermarkets_in_berlin",
    con=engine,
    schema='test_berlin_data',
    if_exists="append",   # or "replace" or "fail" — use "append" in test phase
    index=False
)

355

**🧪 3. Optional — Validate Test Insert**

In [11]:
# Check if data was inserted
with engine.connect() as conn:
    count = conn.execute(text("SELECT COUNT(*) FROM supermarkets_in_berlin")).scalar()
    print(f"✅ {count} rows in the DB.")

✅ 1355 rows in the DB.


In [12]:
from sqlalchemy import create_engine, text

# Define your database connection string
DATABASE_URL = "postgresql+psycopg2://neondb_owner:a9Am7Yy5r9_T7h4OF2GN@ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech:5432/neondb?sslmode=require"
engine = create_engine(DATABASE_URL)

# Define the SQL to get primary key columns
get_primary_key_sql = """
SELECT 
    a.attname AS column_name
FROM 
    pg_index i
JOIN 
    pg_attribute a ON a.attrelid = i.indrelid
                  AND a.attnum = ANY(i.indkey)
WHERE 
    i.indrelid = 'supermarkets_in_berlin'::regclass
    AND i.indisprimary;
"""

# Execute and print results
with engine.connect() as conn:
    result = conn.execute(text(get_primary_key_sql)).fetchall()
    primary_keys = [row[0] for row in result]
    print("🗝️ Primary key columns:", primary_keys)

🗝️ Primary key columns: []


# Step 4: Populate dataset into production database DBeaver-PostgreSQL

In [1]:
# Libraries
import psycopg2
import pandas as pd
from sqlalchemy import create_engine, text
import warnings

warnings.filterwarnings("ignore")

In [8]:
df_final= pd.read_csv("../sources/final_supermarkets_with_district.csv")

In [7]:
user_name='thofa_tazkia'
password='775oIqBN7KlPlzQG'

In [6]:
# Conection
host = 'localhost'
port = '5433'
database = 'layereddb'
schema_='berlin_source_data'

# tunnel is already created
#connection to db after you opened tunnel
engine_ = create_engine(f'postgresql+psycopg2://{user_name}:{password}@{host}:{port}/{database}')

In [12]:
#creating table with constraints and references first

create_table_query = f"""
CREATE TABLE IF NOT EXISTS {schema_}.supermarkets(
    store_id TEXT PRIMARY KEY,
    store_name TEXT,
    street TEXT,
    housenumber TEXT,
    postcode TEXT,
    city TEXT,
    opening_hours TEXT,
    brand TEXT,
    type TEXT,
    payment_credit_card TEXT,
    payment_debit_cards TEXT,
    payment_cash TEXT,
    payment_contactless TEXT,
    wheelchair TEXT,
    internet_access TEXT,
    phone TEXT,
    email TEXT,
    website TEXT,
    latitude DOUBLE PRECISION,
    longitude DOUBLE PRECISION,
    source TEXT,
    district TEXT,
    neighborhood TEXT,
    neighborhood_id TEXT,
    district_id TEXT,
    CONSTRAINT district_id_fk FOREIGN KEY (district_id)
        REFERENCES berlin_source_data.districts(district_id)
        ON DELETE RESTRICT
        ON UPDATE CASCADE
        
);
"""
# Execute the query to create empty table
with engine_.connect() as conn:
     conn.execute(text(create_table_query))
     conn.commit()
     print("✅ Table created successfully")

✅ Table created successfully


In [13]:
#  Send the DataFrame to the database using .to_sql()
df_final.to_sql(
    "supermarkets",
    engine_,
    schema="berlin_source_data",
    if_exists="append",   # append to existing
    index=False
)

355

In [14]:
##let's query test data!
query = f"""
SELECT * from berlin_source_data.districts
"""

# Execute the query
with engine_.connect() as conn:
    df= pd.read_sql(text(query), conn)
    conn.commit()  # commit the transaction
df.head(3)

district_id                    district  \
0    11012012               Reinickendorf   
1    11004004  Charlottenburg-Wilmersdorf   
2    11009009            Treptow-Köpenick   

                                            geometry  
0  0106000020E61000000100000001030000000100000084...  
1  0106000020E6100000010000000103000000010000000D...  
2  0106000020E610000001000000010300000001000000E9...

In [15]:
##let's query test data!
queryy = f"""
SELECT * from berlin_source_data.supermarkets
"""

# Execute the query
with engine_.connect() as conn:
    df= pd.read_sql(text(queryy), conn)
    conn.commit()  # commit the transaction
df.head(3)

store_id             store_name                  street housenumber  \
0  58489979  netto marken-discount        alte jakobstraße          83   
1  79418658                   ledo            quäkerstraße           2   
2  79422426              kiezmarkt  Auguste-Viktoria-Allee     unknown   

  postcode    city                      opening_hours                  brand  \
0    10179  berlin          mo-sa 07:00-24:00; su off  netto marken-discount   
1    13403  berlin                  mo-sa 09:00-20:00            independent   
2    13403  Berlin  mo-fr 08:00-20:00, sa 08:00-19:00                nahkauf   

      type payment_credit_card  ...         phone         email  \
0  unknown             unknown  ...  not_provided  not_provided   
1  unknown             unknown  ...  not_provided  not_provided   
2  unknown             unknown  ...  not_provided  not_provided   

                                             website   latitude  longitude  \
0  https://www.netto-online.de/filialen/berlin-mi...  52.509819  13.407373   
1                    https://www.ledo-supermarkt.de/  52.572417  13.308547   
2                                       not_provided  52.571676  13.312476   

  source       district   neighborhood  neighborhood_id  district_id  
0    osm          mitte          mitte              101     11001001  
1    osm  reinickendorf  reinickendorf             1201     11012012  
2    osm  reinickendorf  reinickendorf             1201     11012012  

[3 rows x 25 columns]

In [16]:
from sqlalchemy import text

# Define the SQL query to check the foreign key constraints
check_fk_query = """
SELECT
    conname AS constraint_name,
    conrelid::regclass AS table_name,
    a.attname AS column_name,
    confrelid::regclass AS foreign_table_name
FROM
    pg_constraint AS c
JOIN
    pg_attribute AS a ON a.attnum = ANY(c.conkey)
    AND a.attrelid = c.conrelid
WHERE
    c.contype = 'f'
    AND conrelid = 'berlin_source_data.supermarkets'::regclass;
"""

# Execute and print result
with engine_.connect() as conn:
    result = conn.execute(text(check_fk_query)).fetchall()
    print("🔗 Foreign keys:")
    for row in result:
        print(dict(row._mapping))

🔗 Foreign keys:
{'constraint_name': 'district_id_fk', 'table_name': 'berlin_source_data.supermarkets', 'column_name': 'district_id', 'foreign_table_name': 'berlin_source_data.districts'}
